witchcraft-server/src/service/trace_propagation.rs

```diff
@@ -91,20 +91,29 @@ where
         span.tag("http.version", &format!("{:?}", req.version()));
 
         TracePropagationFuture {
-            inner: self.inner.call(req),
+            inner: Some(self.inner.call(req)),
             span: Some(span),
         }
         .await
     }
 }
 
-#[pin_project]
+#[pin_project(PinnedDrop)]
 pub struct TracePropagationFuture<F> {
     #[pin]
-    inner: F,
+    inner: Option<F>,
     span: Option<OpenSpan<Detached>>,
 }
 
+#[pinned_drop]
+impl<F> PinnedDrop for TracePropagationFuture<F> {
```

Comment:
These are required to ensure that the zipkin thread local state is set when the endpoint metric layer handles its timer updates.